README.md
# flask-healthcheck
Flask-Healthcheck is a [Flask](http://flask.pocoo.org/) extension that provides
a simple way of adding healthchecks to a service.

## Example:
```python
from flask import Flask
app = Flask(__name__)
healthcheck = Healthcheck(app)

@app.route('/')
def hello_world():
    return 'Hello world!'


@healthcheck
def connections():
    if len(connections) < 5:
        return False, 'Less than 5 active connections'
    else:
        return True
```

Making a request against this endoint will return a 200 status code and the
following output when the healthcheck passes:
```json
{
  "connections": {
    "healthy": true,
  }
}
```

And, the following if the healthcheck fails:
```json
{
  "connections": {
    "healthy": false,
    "message": "Less than 5 active connections"
  }
}
```

## Built-in Healthchecks
Healthchecks for some popular Flask extensions are already built in and just
need to be activated:
```python
from flask_sqlalchemy import SQLAlchemy
db = SQLAlchemy(app)
healthcheck.add_extension('sqlalchemy', db)
```
Here it's important to call `add_extension()` with the proper extension name
and the extension object.



debian/changelog
flask-healthcheck (0.1.2) unstable; urgency=low

  * Initial release

 -- Thomas Sarboni <max-k@post.com>  Thu, 03 Aug 2017 10:52:00 +0100


flask_healthcheck/__init__.py
import functools

from flask import current_app, jsonify

from .extensions import registered_extensions


class Healthcheck:

    def __init__(self, app=None):
        if app:
            self.init_app(app)
        self.healthchecks = {}

    def __call__(self, healthcheck_func, name=None):
        if not name:
            name = healthcheck_func.__name__
        self.healthchecks[name] = healthcheck_func
        return healthcheck_func

    def init_app(self, app):
        rule = app.config.get('HEALTHCHECK_PATH', '/healthcheck')
        app.add_url_rule(rule, view_func=self.do_healthcheck)

    def do_healthcheck(self):
        response = {}
        ok = True

        for name, healthcheck in self.healthchecks.items():
            try:
                result = healthcheck()
                if isinstance(result, tuple):
                    state, message = result
                else:
                    state, message = result, None
            except Exception as e:
                state = False
                message = str(e)
            response[name] = {'healthy': state}
            if message:
                response[name]['message'] = message
            ok &= state
        return jsonify(response), 200 if ok else 500

    def add_extension(self, ext_name, ext_obj):
        if ext_name not in registered_extensions:
            raise ValueError("No built in healthcheck for extension '{}'"
                             .format(ext_name))
        name, func = registered_extensions[ext_name]
        self.healthchecks[name] = functools.partial(func, ext_obj)



flask_healthcheck/extensions.py
registered_extensions = {}


def register_extension(healthcheck_name=None):
    def decorator(healthcheck_func):
        ext_name = healthcheck_func.__name__
        registered_extensions[ext_name] = (healthcheck_name, healthcheck_func)
        return healthcheck_func

    # If called without parameters
    if callable(healthcheck_name):
        healthcheck_func = healthcheck_name
        healthcheck_name = healthcheck_func.__name__
        decorator = decorator(healthcheck_func)
    return decorator


@register_extension('sql')
def sqlalchemy(db):
    with db.engine.connect() as connection:
        connection.execute('SELECT 1;')
    return True


@register_extension('mongodb')
def pymongo(mongo):
    mongo.db.command('ping')
    return True


@register_extension('mongodb')
def mongoengine(mongo):
    mongo.connection.command('ping')
    return True


@register_extension
def redis(db):
    db.ping()
    return True



flask_healthcheck.py
import functools

from flask import current_app, jsonify


class Healthcheck:
    registered_extensions = {}

    def __init__(self, app=None):
        if app:
            self.init_app(app)
        self.healthchecks = {}

    def __call__(self, healthcheck_func, name=None):
        if not name:
            name = healthcheck_func.__name__
        self.healthchecks[name] = healthcheck_func
        return healthcheck_func

    def init_app(self, app):
        rule = app.config.get('HEALTHCHECK_PATH', '/healthcheck')
        app.add_url_rule(rule, view_func=self.do_healthcheck)

    def do_healthcheck(self):
        response = {}
        ok = True

        for name, healthcheck in self.healthchecks.items():
            try:
                result = healthcheck()
                if isinstance(result, tuple):
                    state, message = result
                else:
                    state, message = result, None
            except Exception as e:
                state = False
                message = str(e)
            response[name] = {'healthy': state}
            if message:
                response[name]['message'] = message
            ok &= state
        return jsonify(response), 200 if ok else 500

    @classmethod
    def register_extension(cls, healthcheck_name=None):
        def decorator(healthcheck_func):
            ext_name = healthcheck_func.__name__
            cls.registered_extensions[ext_name] = (healthcheck_name, healthcheck_func)
            return healthcheck_func

        # If called without parameters
        if callable(healthcheck_name):
            healthcheck_func = healthcheck_name
            healthcheck_name = healthcheck_func.__name__
            decorator = decorator(healthcheck_func)
        return decorator

    def add_extension(self, ext_name, ext_obj):
        if ext_name not in self.registered_extensions:
            raise ValueError("No built in healthcheck for extension '{}'"
                             .format(ext_name))
        name, func = self.registered_extensions[ext_name]
        self.healthchecks[name] = functools.partial(func, ext_obj)


@Healthcheck.register_extension('sql')
def sqlalchemy(db):
    with db.engine.connect() as connection:
        connection.execute('SELECT 1;')
    return True


@Healthcheck.register_extension('mongodb')
def pymongo(mongo):
    mongo.db.command('ping')
    return True


@Healthcheck.register_extension('mongodb')
def mongoengine(mongo):
    mongo.connection.command('ping')
    return True


@Healthcheck.register_extension
def redis(db):
    db.ping()
    return True

